libp2p/protocols/pubsub/pubsubpeer.nim

```diff
@@ -144,21 +151,36 @@ proc send*(p: PubSubPeer, msg: RPCMsg) {.async.} =
     libp2p_pubsub_skipped_sent_messages.inc(labelValues = [p.id])
     return
 
-  try:
+  proc sendToRemote() {.async.} =
+    logScope:
+      peer = p.id
+      msg = shortLog(msg)
+
     trace "about to send message"
+
+    if not p.onConnect.isSet:
+      await p.onConnect.wait()
+
     if p.connected: # this can happen if the remote disconnected
       trace "sending encoded msgs to peer"
 
       await p.sendConn.writeLp(encoded)
       p.sentRpcCache.put(digest)
+      trace "sent pubsub message to remote"
 
       for x in mm.messages:
         for t in x.topicIDs:
           # metrics
           libp2p_pubsub_sent_messages.inc(labelValues = [p.id, t])
 
+  let sendFut = sendToRemote()
+  try:
+    await sendFut.wait(timeout)
```

Comment:
this is where libp2p hangs in linux @sinkingsugar @cheatfate @stefantalpalaru @arnetheduck

Reply:
specifically, the `await` in tha produced introduces the hangs, changing to `asyncCheck` fixes finalization on linux.